smalldata_tools/ana_funcs/droplet2Photons.py

```diff
@@ -8,71 +8,256 @@
-        self.threshold = kwargs.get('threshold', None)
-        self.mask = kwargs.get('mask', None)
         self.aduspphot = kwargs.get('aduspphot', 0)
-        self.offset = kwargs.get('offset', 0)
+        self.offset = kwargs.get('offset', self.aduspphot*0.5)
         self.photpts = np.arange(1000000)*self.aduspphot-self.aduspphot+self.offset
+        self.one_photon_info = kwargs.get('one_photon_info', False)
         # self.Np = kwargs.get('Np', None)
         self.cputime = kwargs.get('cputime', False)
-        
-    def setFromDet(self, det):
-        super(droplet2Photons, self).setFromDet(det)
-        if self.mask is None:
-            self.mask = det.mask
+        self._footprintnbr = np.array([[0,1,0],[1,0,1],[0,1,0]])
+        self.one_photon_limits = kwargs.get('one_photon_limits', None)
+        # not sure if this should not just be photpts[:2]?
+        if self.one_photon_limits is None:
+            self.one_photon_limits = [self.offset, self.offset+self.aduspphot]
+
+    @jit(nopython=True, cache=True)
+    def piximg(self, i,j,adu, pad=True):
+        """
+        make a sub-image of only the droplet
+        takes coordinates & values, equivalent to sparse matrix w/ optionial edge pixels
+        """
+        if pad:
+            img = np.zeros((np.max(i)-np.min(i)+3,np.max(j)-np.min(j)+3))
+            zip_obj_old = zip(i+1,j+1,adus)
         else:
-            self.mask = np.logical_and(self.mask, det.mask)
-        
+            img = np.zeros((np.max(i)-np.min(i)+1,np.max(j)-np.min(j)+1))
+            zip_obj_old = zip(i,j,adus)
+        zip_obj = NTL()
+        [zip_obj.append(x) for x in zip_obj_old]
+
+        mi = np.min(i)
+        mj = np.min(j)
+        for ti,tj,tadu in zip_obj:
+            img[ti-mi,tj-mj]=tadu
+
+#    @jit(nopython=True, cache=True)
+    def onephoton_max(self, pixones, npix_drop, ppos):
+        """
+        pixones: adus/droplet
+        npix_drop: number of pixels/droplet
+        ppos: array of pixel positons of droplets
+        returns arrays of the energy of the maximum pixel & highest two pixels for 1-photon droplets
+        """
+        maxPixAdu=[]
+        twoPixAdu=[]
+        twobnrPixAdu=[]
+        for drop in range(len(npix_drop)):
+            i = pixones[ppos[drop]:ppos[drop+1],0]
+            j = pixones[ppos[drop]:ppos[drop+1],1]
+            adus = pixonesadu[ppos[drop]:ppos[drop+1]]
+            maxPixAdu.append(np.nanmax(adus))
+            ##they need to be neighbors, this is not required here....
+            #twoPixAdu.append((np.sort(adus)[-1*(min(2, len(adus))):]).sum())
+            piximg1=sparse.coo_matrix( (adus, (i, j)) ).todense()
+            #find maximum, find maximum of footprint_nbradd neighbor, maximum.
+            nbrpix=filters.maximum_filter(piximg1, mode='constant', footprint=self._footprintnbr).flatten()[np.argmax(piximg)]
+            twonbrPixAdu.append(maxPixAdu[-1]+nbrpix)
+            return maxPixAdu, twonbrPixAdu
```

Comment:
Should we indent the return statement back?

Reply:
Done